app/controllers/admin/sponsors_controller.rb

```diff
@@ -19,38 +22,51 @@ def create
     @sponsor = Sponsor.new(sponsor_params)
     @sponsor.build_address if @sponsor.address.blank?
     authorize @sponsor
-    if @sponsor.save
+
+    if @sponsor.valid? && @sponsor.save
       flash[:notice] = "Sponsor #{@sponsor.name} created"
-      redirect_to [:admin, @sponsor]
-    else
-      flash[:notice] = @sponsor.errors.full_messages.to_s
-      render 'new'
+      return redirect_to [:admin, @sponsor]
     end
+
+    flash[:warning] = @sponsor.errors.full_messages
+    render 'new'
   end
 
   def edit
     @sponsor = Sponsor.find(params[:id])
     authorize @sponsor
     @sponsor.build_address if @sponsor.address.blank?
-    @sponsor.contacts.build
+    @sponsor.contacts.build unless @sponsor.contacts.any?
   end
 
   def update
-    @sponsor.update(sponsor_params)
-    redirect_to admin_sponsor_path(@sponsor), notice: 'Updated!'
+    @sponsor.assign_attributes(sponsor_params)
+
+    if @sponsor.valid? && @sponsor.save
+      flash[:notice] = 'Updated!'
+      return redirect_to admin_sponsor_path(@sponsor)
+    end
+
+    flash[:warning] = @sponsor.errors.full_messages
+    render 'edit'
   end
 
   private
 
   def sponsor_params
     params.require(:sponsor).permit(:name, :avatar, :website, :seats, :accessibility_info,
                                     :number_of_coaches, :level, :email, :contact_first_name,
-                                    :contact_surname, contact_ids: [], address_attributes:
```

Comment:
@despo am I correct in thinking the `contact_first_name` & `contact_surname` fields are going to be removed/have been removed from Sponsor in another PR? I think you mentioned something to this effect in the PR description but wanted to double check.

Reply:
Yes as I want to run the migrations first. However I should have removed them from the permitted params now, thanks for pointing that out!